src/jdk.jlink/share/classes/jdk/tools/jlink/internal/plugins/SystemModulesPlugin.java

```diff
@@ -1580,237 +1034,195 @@ void hashForModule(String name, byte[] hash) {
-             */
-            int indexOfExportsModifiers(CodeBuilder cob, Set<Exports.Modifier> mods) {
-                return exportsModifiersSets.get(mods).build(cob);
+            private <T extends Enum<T>> Snippet buildEnumSet(ClassBuilder clb, SetReference<T> setRef) {
+                return cacheBuilder.transform(setRef,
+                        new SetSnippetBuilder(CD_Object)
+                                .classBuilder(clb)
+                                .activatePagingThreshold(PAGING_THRESHOLD)
+                                .ownerClassDesc(owner)
+                                .methodNamePrefix("dedupSet" + setBuilt++)
+                                .pageSize(ENUM_PAGE_SIZE)
+                                .build(Snippet.buildAll(setRef.elements(), Snippet::loadEnum)));
             }
 
-            /**
-             * Retrieve the index to the given set of Opens.Modifier.
-             * Emit code to generate it when EnumSetBuilder::build is called.
-             */
-            int indexOfOpensModifiers(CodeBuilder cob, Set<Opens.Modifier> mods) {
-                return opensModifiersSets.get(mods).build(cob);
+            private Map<Set<String>, Snippet> buildStringSets(ClassBuilder clb, Map<Set<String>, SetReference<String>> map) {
+                Map<Set<String>, Snippet> snippets = new HashMap<>(map.size());
+                map.entrySet().forEach(entry -> snippets.put(entry.getKey(),
+                        buildStringSet(clb, entry.getValue())));
+                return snippets;
             }
 
+            private <T extends Enum<T>> Map<Set<T>, Snippet> buildEnumSets(ClassBuilder clb, Map<Set<T>, SetReference<T>> map) {
+                Map<Set<T>, Snippet> snippets = new HashMap<>(map.size());
+                map.entrySet().forEach(entry -> snippets.put(entry.getKey(),
+                        buildEnumSet(clb, entry.getValue())));
+                return snippets;
+            }
 
-            /*
-             * Retrieve the index to the given set of Requires.Modifier.
-             * Emit code to generate it when EnumSetBuilder::build is called.
-             */
-            int indexOfRequiresModifiers(CodeBuilder cob, Set<Requires.Modifier> mods) {
-                return requiresModifiersSets.get(mods).build(cob);
+            DedupSet build(ClassBuilder clb) {
+                return new DedupSet(
+                    buildStringSets(clb, stringSets),
+                    buildEnumSets(clb, requiresModifiersSets),
+                    buildEnumSets(clb, opensModifiersSets),
+                    buildEnumSets(clb, exportsModifiersSets),
+                    cacheBuilder.build(clb)
+                );
             }
 
             /*
-             * Build a new string set without any attempt to deduplicate it.
+             * SetReference count references to the set, and use LoadableSet under the hood to
+             * support paginiation as needed.
```

Comment:
```suggestion
             * support pagination as needed.
```

Reply:
The code has been refactored.  The comment should be updated to reflect the new code?

Reply:
Will do a pass in next commit, I want to see if this refactoring favorable.

Reply:
Yes it does.   Thanks for the update.